core/src/main/java/io/grpc/internal/TransportTracer.java

```diff
@@ -0,0 +1,180 @@
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A class for gathering statistics about a transport. This is an experimental feature.
+ * Can only be called from the transport thread unless otherwise noted.
+ */
+public final class TransportTracer {
+  private long streamsStarted;
+  private long lastStreamCreatedTimeNanos;
+  private long streamsSucceeded;
+  private long streamsFailed;
+  private long keepAlivesSent;
+  private FlowControlReader flowControlWindowReader;
+
+  // TODO(zpencer): msg sent can piggyback on framer's writequeue and avoid syncing the update
+  private final LongCounter messagesSent = LongCounterFactory.create();
+  private volatile long lastMessageSentTimeNanos;
+  // deframing happens on the application thread, and there's no easy way to avoid synchronization
+  private final LongCounter messagesReceived = LongCounterFactory.create();
+  private volatile long lastMessageReceivedTimeNanos;
+
+  /**
+   * Returns a read only set of current stats.
+   */
+  public Stats getStats() {
+    return new Stats(
+        streamsStarted,
+        lastStreamCreatedTimeNanos,
+        streamsSucceeded,
+        streamsFailed,
+        messagesSent.value(),
+        messagesReceived.value(),
+        keepAlivesSent,
+        lastMessageSentTimeNanos,
+        lastMessageReceivedTimeNanos,
+        flowControlWindowReader);
+  }
+
+  /**
+   * Called by the transport to report a stream has started. For clients, this happens when a header
+   * is sent. For servers, this happens when a header is received.
+   */
+  public void reportStreamStarted() {
+    streamsStarted++;
+    lastStreamCreatedTimeNanos = currentTimeNanos();
+  }
+
+  /**
+   * Reports that a stream closed with the specified Status.
+   */
+  public void reportStreamClosed(Status status) {
+    if (status.isOk()) {
+      streamsSucceeded++;
+    } else {
+      streamsFailed++;
+    }
+  }
+
+  /**
+   * Reports that a message was successfully sent. This method is thread safe.
+   */
+  public void reportMessageSent() {
+    messagesSent.add(1);
+    lastMessageSentTimeNanos = currentTimeNanos();
+  }
+
+  /**
+   * Reports that a message was successfully received. This method is thread safe.
+   */
+  public void reportMessageReceived() {
+    messagesReceived.add(1);
+    lastMessageReceivedTimeNanos = currentTimeNanos();
+  }
+
+  /**
+   * Reports that a keep alive message was sent.
+   */
+  public void reportKeepAliveSent() {
+    keepAlivesSent++;
+  }
+
+  /**
+   * Registers a {@link FlowControlReader} that can be used to read the local and remote flow
+   * control window sizes.
+   */
+  public void setFlowControlWindowReader(FlowControlReader flowControlWindowReader) {
+    this.flowControlWindowReader = Preconditions.checkNotNull(flowControlWindowReader);
+  }
+
+  /**
+   * A container that holds the local and remote flow control window sizes.
+   */
+  public static final class FlowControlWindows {
+    public final int remoteBytes;
```

Comment:
nit: these should probably be longs

Reply:
Netty returns only `int`s here, but I don't have a problem upcasting it in case that changes.

Reply:
Update: The http/2 spec states that the flow control window must always be below 2^31-1, which makes an int technically sufficient.

```
A sender MUST NOT allow a flow-control window to exceed 2^31-1 octets. If a sender receives a WINDOW_UPDATE that causes a flow-control window to exceed this maximum, it MUST terminate either the stream or the connection, as appropriate. For streams, the sender sends a RST_STREAM with an error code of FLOW_CONTROL_ERROR; for the connection, a GOAWAY frame with an error code of FLOW_CONTROL_ERROR is sent.
```